# Introduction to Notebooks

*Tyler Caraza-Harter*


In this semester, we've learned two ways to run Python code: in interactive mode, where code runs as we type it, and script mode, where we write our code in a .py file, which we then run.

Now, we'll going to learn a third way: with Notebooks.  Notebooks are a popular tool of data scientists for interleaving code with analysis results (e.g., tables and plots).  You're going to need to install Anaconda for this part of the course (if you haven't already).

To get started, run the following in the terminal (make sure to use the "cd" command to move to a directory where you want to do your work, such as your documents directory):

```
ipython notebook
```

If things are working properly, you'll see something like the following in the terminal:

```
ty-mac:lec-26$ ipython notebook
[TerminalIPythonApp] WARNING | Subcommand `ipython notebook` is deprecated and will be removed in future versions.
[TerminalIPythonApp] WARNING | You likely want to use `jupyter notebook` in the future
[I 10:36:31.799 NotebookApp] JupyterLab beta preview extension loaded from /anaconda3/lib/python3.6/site-packages/jupyterlab
[I 10:36:31.799 NotebookApp] JupyterLab application directory is /anaconda3/share/jupyter/lab
[I 10:36:31.808 NotebookApp] Serving notebooks from local directory: /Users/trh/git_co/caraza-harter-com/tyler/cs301/fall18/materials/readings/lec-26
[I 10:36:31.808 NotebookApp] 0 active kernels
[I 10:36:31.808 NotebookApp] The Jupyter Notebook is running at:
[I 10:36:31.808 NotebookApp] http://localhost:8888/?token=1254607604289711443154541
```

Also, your web browser should pop up a new page that says Jupyter in the top left.  Jupyter Notebooks are an interactive tool for doing data science in Julia, Python, and R (we'll only be working with Python).

Click the "New" button, then choose "Python 3".  A new document should open, with a box labeled "In [  ]" in which you can start typing.  From here, you can write documentation and Python code, summarize data, and display plots.  In fact, this document your reading now is itself a Python Notebook, and it will include a mix of documentation and code.

## Similarities to Interactive Mode

We can type Python expressions in these boxes
and see the result in an "Out[N]" box.
For example:

In [1]:
1 + 2

3

This should feel similar to interactive Python, but one difference is that we only see the output from the last line.  For example, in the following example, we do two computations, but only see the second result.

In [2]:
3*4 # expression 1

10*10 # expression 2

100

Of course, we can still always print things:

In [3]:
print(3*4)
print(10*10)

12
100


## Ordering Pitfalls

One difference between an ipython notebook and interactive mode is that you can go back and edit earlier cells (each "In [N]" box is called a cell).  While this is useful, there's one gotcha you need to be aware of.

In interactive mode, this code would fail:

```
>>> print(x)
>>> x = 3
```

Similarly, if you try to assign x in a later cell and print it in an earlier cell, it should (and usually does) fail.  One one catch is that notebooks don't force you to execute each cell in order.  You could do something like this:

EARLIER CELL:
print("hi")

LATER CELL:
x = 3

...Run everything...

Modify EARLIER cell:

EARLIER CELL:
print(x)

Run just the EARLIER cell.  Surprisingly, (and deceptively), this works, because you ran the cells out of order.  However, you share you notebook with somebody else, it will fail for them because they might try rerunning everything from scratch, in order.

**Important**: to make sure you don't make this common blunder, be sure to occasionally re-run all your code from the start by clicking "Kernel" from the notebook menu and selecting "Restart & Run All".


# Pandas

A common module you'll use when working in a notebook is called "pandas".  Pandas will allow you to do many of the things you might do in Excel (or another spreadsheet program) programmatically with Python.  Let's import it.

In [4]:
import pandas as pd

The "as pd" expression may be new for you.  This just gives the pandas module a new name in our code, so we can type things like `pd.some_function()` to call a function named `some_function` rather than type out `pandas.some_function`.  You could also have just used `import pandas` or even given it another name with an import like `import pandas as silly_bear`, but we recommend you import as "pd", because most pandas users do so by convention.

We'll also be using to new data types (Series and DataFrame) from Pandas very often, so let's import these directly so we don't even need to prefix their use with "pd.".

In [5]:
from pandas import Series, DataFrame

# Series

A Series is very similar to a Python list, and we can convert back and forth between them:

In [6]:
num_list = [100, 200, 300]
print(type(num_list))

num_series = Series(num_list) # create Series from list
print(type(num_series))

<class 'list'>
<class 'pandas.core.series.Series'>


In [7]:
# displaying a list:
num_list

[100, 200, 300]

In [8]:
# displaying a Series:
num_series

0    100
1    200
2    300
dtype: int64

Notice that both the list and the Series contain the same values.  However, there are some differences:
* the Series is displayed vertically
* the indexes for the series are explicitly displayed by the values
* at the end, it say "dtype: int64"

"dtype" stands for data type.  In this case, it means that the series contains integers, each of which require 64 bits of memory (this detail is not important for us).  Although you could create a Series containing different types of data (as with lists), we'll avoid doing so because working with Series of one type will often be more convenient.

Going from a Series back to a list is just as easy as going from a list to a Series:

In [9]:
list(num_series)

[100, 200, 300]

## Element-Wise Operations


With Series, it is easy to apply the same operation to every value in the Series with a single line of code (instead of a loop).

For example, suppose we wanted to add 1 to every item in a list.  We would need to write something like this:

In [10]:
orig_nums = [100, 200, 300]
new_nums = []
for x in orig_nums:
    new_nums.append(x+1)
new_nums

[101, 201, 301]

With a Series, we can do the same like this:

In [11]:
nums = Series([100, 200, 300])
nums + 1

0    101
1    201
2    301
dtype: int64

This means multiplication means something very different for lists than for Series.

In [12]:
[1,2,3] * 3

[1, 2, 3, 1, 2, 3, 1, 2, 3]

In [13]:
Series([1,2,3]) * 3

0    3
1    6
2    9
dtype: int64

Whereas a "+" means concatenate for lists, it means element-wise addition for Series:

In [14]:
[10, 20] + [3, 4]

[10, 20, 3, 4]

In [15]:
Series([10, 20]) + Series([3, 4])

0    13
1    24
dtype: int64

One implication of this is that you might not get what you expect if you add Series of different sizes:

In [16]:
Series([10,20,30]) + Series([1,2])

0    11.0
1    22.0
2     NaN
dtype: float64

The 10 gets added with the 1, and the 20 gets added with the 2, but there's nothing in the second series to add with 30.  30 plus nothing doesn't make sense, so Pandas gives "NaN".  This stands for "Not a Number".

## Boolean Element-Wise Operation

Consider the following:

In [17]:
nums = Series([1,9,8,2])
nums

0    1
1    9
2    8
3    2
dtype: int64

In [18]:
nums > 5

0    False
1     True
2     True
3    False
dtype: bool

This example shows that you can do element-wise comparisons as well.  The result is a Series of booleans.  If the value in the original Series is greater than 5, we see True at the same position in the output Series.  Otherwise, the value at the same position in the output Series is False.

We can also chain these operations together:

In [19]:
nums = Series([7,5,8,2,3])
nums

0    7
1    5
2    8
3    2
4    3
dtype: int64

In [20]:
mod_2 = nums % 2
mod_2

0    1
1    1
2    0
3    0
4    1
dtype: int64

In [21]:
odd = mod_2 == 1
odd

0     True
1     True
2    False
3    False
4     True
dtype: bool

As you can see, we first obtained an integer Series (`mod_2`) by computing the value of every number modulo 2 (`mod_2`) will of course contain only 1's and 0's).

We then create a Boolean series (`odd`) by comparing the `mod_2` series to 1.

If a number in the `nums` Series is odd, then the value at the same position in the `odd` series will be True.

## Data Alignment

Notice what happens when we create a series from a list:

In [22]:
Series([100,200,300])

0    100
1    200
2    300
dtype: int64

We see the following:
* the first position has index 0 and value 100
* the second position has index 1 and value 200
* the third position has index 2 and value 300

One interesting difference between lists and Series is that with Series, the index does not always need to correspond so closely with the position; that's just a default that can be overridden.

For example:

In [23]:
nums1 = Series([100, 200, 300], index=[2,1,0])
nums1

2    100
1    200
0    300
dtype: int64

Now we see indexes are assigned based on the argument we passed for index (not the position):
* the first position has index 2 and value 100
* the second position has index 1 and value 200
* the third position has index 0 and value 300

When we do element-wise operations between two Sersies, Pandas lines up the data based on index, not position.  As a concrete example, consider three Series:

In [24]:
X = Series([100, 200, 300])
Y = Series([10, 20, 30])
Z = Series([10, 20, 30], index=[2,1,0])

In [25]:
X

0    100
1    200
2    300
dtype: int64

In [26]:
Y

0    10
1    20
2    30
dtype: int64

In [27]:
Z

2    10
1    20
0    30
dtype: int64

**Note:** Y and Z are nearly the same (numbers 10, 20, and 30, in that order), except for the index.  Let's see the difference between `X+Y` and `Y+Z`:

In [28]:
X+Y

0    110
1    220
2    330
dtype: int64

In [29]:
X+Z

0    130
1    220
2    310
dtype: int64

For `X+Y`, Pandas adds the number at index 0 in X (100) with the value at index 0 in Y (10), such that the value in the output at index 0 is 110.

For `X+Z`, Pandas adds the number at index 0 in X (100) with the value at index 0 in Y (30), such that the value in the output at index 0 is 130.  It doesn't matter that the first number in Z is 10, because Pandas does element-wise operations based on index, not position.

## Series Indexing and Slicing

We will often one to extract one or more values from a Series.  Like lists, Series support indexing and slicing.

In [32]:
letter_list = ["A", "B", "C", "D"]
letter_series = Series(letter_list)
letter_series

0    A
1    B
2    C
3    D
dtype: object

In [33]:
letter_list[0]

'A'

In [34]:
letter_series[0]

'A'

In [35]:
letter_list[3]

'D'

In [36]:
letter_series[3]

'D'

In [37]:
letter_list[-1]

'D'

In [38]:
# but be careful!  Series don't support negative indexes like lists do
try:
    print(letter_series[-1])
except Exception as e:
    print(type(e))


<class 'KeyError'>


Series slicing works much like list slicing:

In [44]:
print("list slice:")
print(letter_list[:2])
print("\nseries slice:")
print(letter_series[:2])

list slice:
['A', 'B']

series slice:
0    A
1    B
dtype: object


In [45]:
print("list slice:")
print(letter_list[2:])
print("\nseries slice:")
print(letter_series[2:])

list slice:
['C', 'D']

series slice:
2    C
3    D
dtype: object


In [47]:
# although we CANNOT do negative indexing with a Series
# we CAN use negative numbers in a Series slice
print("list slice:")
print(letter_list[:-1])
print("\nseries slice:")
print(letter_series[:-1])

list slice:
['A', 'B', 'C']

series slice:
0    A
1    B
2    C
dtype: object


## Series Selectors

In addition to supporting indexing and slicing, Series allow you to pull out items using a Boolean series.  For example:

In [50]:
letters = Series(["A", "B", "C", "D"])
letters

0    A
1    B
2    C
3    D
dtype: object

In [51]:
bool_series = Series([True, True, False, False])
bool_series

0     True
1     True
2    False
3    False
dtype: bool

In [53]:
# we can used the bool_series almost like an index
# to pull values out of letters:

letters[bool_series]

0    A
1    B
dtype: object

In [54]:
# We could also create the Boolean Series on the fly:
letters[Series([True, True, False, False])]

0    A
1    B
dtype: object

In [55]:
# Let's grab the last two letterrs:
letters[Series([False, False, True, True])]

2    C
3    D
dtype: object

In [58]:
# Let's grab the first and last (can't do this with a slice):
letters[Series([True, False, False, True])]

0    A
3    D
dtype: object

## Combining Element-Wise Operations with Selection

As we just saw, we can use a Boolean series (let's call it B) to select values from another Series (let's call it S).

A common pattern is to create B by performing operation on S, then using B to select from S.  Let's try doing this to pull all the numbers greater than 5 from a Series.

**Example 1**

In [61]:
# we want to pull out 9 and 8
S = Series([1,9,2,3,8])
S

0    1
1    9
2    2
3    3
4    8
dtype: int64

In [62]:
B = S > 5
B

0    False
1     True
2    False
3    False
4     True
dtype: bool

In [64]:
# this will pull out values from S at index 1 and 4,
# because the values in B at index 1 and 4 are True
S[B]

1    9
4    8
dtype: int64

** Example 2 **

Let's try to pull out all the upper case strings from a series:

In [66]:
words = Series(["APPLE", "boy", "CAT", "dog"])
words

0    APPLE
1      boy
2      CAT
3      dog
dtype: object

In [73]:
# we can use .str.upper() to get upper case version of words
upper_words = words.str.upper()
upper_words

0    APPLE
1      BOY
2      CAT
3      DOG
dtype: object

In [75]:
# B will be True where the original word equals the upper-case version
B = words == upper_words
B

0     True
1    False
2     True
3    False
dtype: bool

In [77]:
# pull out the just words that were orginally uppercase
words[B]

0    APPLE
2      CAT
dtype: object

We have done this example in several steps to illustrate what is happening, but it could have been simplified.  Recall that B is `words == upper_words`.  Thus we could have done this without ever storing a Boolean series in B:

In [78]:
words[words == upper_words]

0    APPLE
2      CAT
dtype: object

Let's simplify one step further (instead of using upper_words, let's paste the expression we used to compute it earlier):

In [79]:
words[words == words.str.upper()]

0    APPLE
2      CAT
dtype: object

**Example 3**

Let's try to pull out all the odd numbers from this Series:

In [81]:
nums = Series([11,12,19,18,15,17])
nums

0    11
1    12
2    19
3    18
4    15
5    17
dtype: int64

`nums % 2` well produce a Series of 1's (for odd numbers) and 0's (for even numbers).  Thus `nums % 2 == 1` produces a Boolean Series of True's (for odd numbers) and False's (for even numbers).  Let's use that Boolean Series to pull out the odd numbers:

In [83]:
nums[nums % 2 == 1]

0    11
2    19
4    15
5    17
dtype: int64